trie/witness_operators.go

```diff
@@ -0,0 +1,373 @@
+		if err := encoder.Encode(o.Nonce); err != nil {
+			return err
+		}
+	}
+
+	if o.Balance.Cmp(emptyBalance) != 0 {
+		bb := o.Balance.Bytes()
+		if err := encodeByteArray(bb, output.WithColumn(ColumnLeafValues)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (o *OperatorLeafAccount) LoadFrom(input io.Reader) error {
+	key, err := decodeByteArray(input)
+	if err != nil {
+		return err
+	}
+	o.Key = keyBytesToNibbles(key)
+
+	flags := make([]byte, 1)
+	_, err = input.Read(flags)
+	if err != nil {
+		return err
+	}
+	o.HasCode = flags[0]&flagCode != 0
+	o.HasStorage = flags[0]&flagStorage != 0
+
+	if flags[0]&flagNonce != 0 {
+		var nonce uint64
+		decoder := codec.NewDecoder(input, &cbor)
+		err = decoder.Decode(&nonce)
+		if err != nil {
+			return err
+		}
+
+		o.Nonce = nonce
+	}
+
+	balance := big.NewInt(0)
+
+	if flags[0]&flagBalance != 0 {
+		var balanceBytes []byte
+		balanceBytes, err = decodeByteArray(input)
+		if err != nil {
+			return err
+		}
+
+		balance.SetBytes(balanceBytes)
+	}
+
+	o.Balance = balance
+
+	return nil
+}
+
+type OperatorCode struct {
+	Code []byte
+}
+
+func (o *OperatorCode) WriteTo(output *WitnessStatsCollector) error {
+	if err := writeOpCode(OpCode, output.WithColumn(ColumnStructure)); err != nil {
+		return err
+	}
+
+	return encodeByteArray(o.Code, output.WithColumn(ColumnCodes))
+}
+
+func (o *OperatorCode) LoadFrom(input io.Reader) error {
+	code, err := decodeByteArray(input)
+	if err != nil {
+		return err
+	}
+	o.Code = code
+	return nil
+}
+
+type OperatorBranch struct {
+	Mask uint32
+}
+
+func (o *OperatorBranch) WriteTo(output *WitnessStatsCollector) error {
+	if err := writeOpCode(OpBranch, output.WithColumn(ColumnStructure)); err != nil {
+		return err
+	}
+	encoder := codec.NewEncoder(output.WithColumn(ColumnStructure), &cbor)
+	return encoder.Encode(o.Mask)
+}
+
+func (o *OperatorBranch) LoadFrom(input io.Reader) error {
+	var mask uint32
+	decoder := codec.NewDecoder(input, &cbor)
+	if err := decoder.Decode(&mask); err != nil {
+		return err
+	}
+
+	o.Mask = mask
+
+	return nil
+}
+
+type OperatorEmptyRoot struct{}
+
+func (o *OperatorEmptyRoot) WriteTo(output *WitnessStatsCollector) error {
+	return writeOpCode(OpEmptyRoot, output.WithColumn(ColumnStructure))
+}
+
+func (o *OperatorEmptyRoot) LoadFrom(input io.Reader) error {
+	// no-op
+	return nil
+}
+
+type OperatorExtension struct {
+	Key []byte
+}
+
+func (o *OperatorExtension) WriteTo(output *WitnessStatsCollector) error {
+	if err := writeOpCode(OpExtension, output.WithColumn(ColumnStructure)); err != nil {
+		return err
+	}
+	return encodeByteArray(keyNibblesToBytes(o.Key), output.WithColumn(ColumnLeafKeys))
+}
+
+func (o *OperatorExtension) LoadFrom(input io.Reader) error {
+	key, err := decodeByteArray(input)
+	if err != nil {
+		return err
+	}
+
+	o.Key = keyBytesToNibbles(key)
+
+	return nil
+}
+
+func decodeByteArray(input io.Reader) ([]byte, error) {
+	codec := codec.NewDecoder(input, &cbor)
```

Comment:
Creating a decoder per operation is expensive. Need a pool or at least decoder per witness